woocommerce/woocommerce
# Move props from datastore to order object and implement $save params for DataStore

### All Submissions:






### Changes proposed in this Pull Request:



We currently store many props in the metadata, and they are set/get directly on the go by the CPT datastore. However, this is problematic because they are technically props of the order, not the datastore. DataStore does not have any representation in the database. 

This causes a problem which is that we have moved some of these props to the custom table. Given that they are directly being manipulated in the post meta table, we end storing duplicates of them (one in the operation table, and then again in the meta (which may be added by the WC core itself with $order->update_meta calls). 

To solve this, we have moved the properties to the WC_Order object itself. This visibility is public, but since datastore was already exposing these props with public visibility, it's the same. This PR also adds a new internal var in the order object called `legacy_datastore_props`; the goal of this prop is that whenever `add/update/delete` meta is called for any of these internal props, we call the respective getter/setter instead, without throwing a warning.

### How to test the changes in this Pull Request:

Adequate unit tests have been added for setting/updating these props, so no additional testing is necessary. However, you can test that no duplicate data is added for these props.

The bug that these props being in metadata would cause is that we would be adding duplicate metadata even though the prop already exists in the table. In this case, props in the operational table would not be updated, but functionality won't be affected since we would add meta rows, and they will be updated as expected. So no visible functionality change can be observed from the UX.

### Other information:




### FOR PR REVIEWER ONLY:
